File: checks/python/src/section508/checks/native_control.py
from section508.dom import attr, closest, has_attr, norm_text, select, tag_name


def check(root):
    if not root:
        return []

    fakes = fake_controls(root)
    if not fakes:
        return []

    sample = norm_text(fakes[0]) or tag_name(fakes[0])
    return [f"“{sample}” is not a native button, link, or input."]


def fake_controls(root):
    fakes = []
    for element in select(root, "div, span, a"):
        if select(element, "button, a[href], input, select, textarea"):
            continue
        if closest(element, "button, a[href], label"):
            continue
        if tag_name(element) == "a":
            href = attr(element, "href")
            if href is None or not href.strip():
                fakes.append(element)
            continue
        if has_attr(element, "onclick"):
            fakes.append(element)
            continue
        if attr(element, "role") and attr(element, "role").lower() == "button":
            fakes.append(element)
    return fakes

File: checks/ruby/lib/section508/checks/native_control.rb
# frozen_string_literal: true

module Section508
  module Checks
    # A clickable job uses button, a[href], or an input.
    class NativeControl
      def self.call(root)
        return [] unless root

        fakes = fake_controls(root)
        return [] if fakes.empty?

        sample = Dom.norm(fakes.first)
        sample = fakes.first.name.downcase if sample.empty?
        ["“#{sample}” is not a native button, link, or input."]
      end

      def self.fake_controls(root)
        return [] unless root.respond_to?(:css)

        Dom.elements(root, "div, span, a").select do |element|
          next false if element.at_css("button, a[href], input, select, textarea")
          next false if Dom.closest(element, "button, a[href], label")

          if element.name.upcase == "A"
            href = element["href"]
            href.nil? || href.strip.empty?
          elsif Dom.has_attr?(element, "onclick")
            true
          else
            element["role"].to_s.downcase == "button"
          end
        end
      end
    end
  end
end
